### PR TITLE
CNS-850: unbond stake entry bug

### PR DESCRIPTION
**Pull request by oren-lava**
In this PR I fix a small bug in which there's a code flow that allowed a provider to fully unbond its stake, but the stake entry is not removed (and there is no unstake entry)

### CODE REVIEW

File: x/dualstaking/keeper/delegate.go [hunk 1]
```diff
@@ -236,7 +237,17 @@ func (k Keeper) decreaseStakeEntryDelegation(ctx sdk.Context, delegator, provide
 		if err != nil {
 			return fmt.Errorf("invalid or insufficient funds: %w", err)
 		}
-		if stakeEntry.Stake.IsLT(k.specKeeper.GetMinStake(ctx, chainID)) {
+		if stakeEntry.Stake.IsZero() {
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegate.go:240]: isnt a dedicated method for this?

Comment (Yaroms) [reply x/dualstaking/keeper/delegate.go:240]: make this a function

**Comment by Yaroms** [comment]
Instead of checking for 0, add a param, min_stake (ulava coin), if the stake is less than that, unstake the provider

File: x/dualstaking/keeper/delegate.go [hunk 2]
```diff
@@ -251,6 +262,27 @@ func (k Keeper) decreaseStakeEntryDelegation(ctx sdk.Context, delegator, provide
 	return nil
 }
 
+func (k Keeper) getUnstakeHoldBlocks(ctx sdk.Context, chainID string) uint64 {
```
Comment (Yaroms) [inline x/dualstaking/keeper/delegate.go:265]: spec should have this function, use it instead

File: x/dualstaking/keeper/migrations.go [hunk 3]
```diff
@@ -375,3 +375,10 @@ func (m Migrator) MigrateVersion3To4(ctx sdk.Context) error {
 	m.keeper.SetDisableDualstakingHook(ctx, false)
 	return nil
 }
+
+// MigrateVersion4To5 sets the MinSelfDelegation param with a default value
+func (m Migrator) MigrateVersion4To5(ctx sdk.Context) error {
+	params := dualstakingtypes.DefaultParams()
```
Comment (Yaroms) [inline x/dualstaking/keeper/migrations.go:381]: set only the required param, dont overwrite everything

Comment (oren-lava) [reply x/dualstaking/keeper/migrations.go:381]: dualstaking had no param before that, so I'm using only the required param

Comment (Yaroms) [reply x/dualstaking/keeper/migrations.go:381]: but you are overwriting all the other params. you should instead:
params := getparams
params.minselfdelegation = types.Defaultparams().minselfdelegation
k.setparams(params)

Comment (Yaroms) [reply x/dualstaking/keeper/migrations.go:381]: ah ok, gotcha

File: x/pairing/keeper/staking.go [hunk 4]
```diff
@@ -182,6 +182,13 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
 		DelegateCommission: delegationCommission,
 	}
```
Comment (omerlavanet) [inline x/pairing/keeper/staking.go:184]: missed to add the same on the flow for an existing entry

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: x/dualstaking/keeper/delegate.go [hunk 5]
```diff
@@ -236,7 +236,28 @@ func (k Keeper) decreaseStakeEntryDelegation(ctx sdk.Context, delegator, provide
...
+				return utils.LavaFormatError("can't remove stake Entry after decreasing provider self delegation", err,
+					utils.Attribute{Key: "index", Value: index},
+					utils.Attribute{Key: "spec", Value: chainID},
+				)
+			}
+			details["min_self_delegation"] = k.GetParams(ctx).MinSelfDelegation.String()
+			utils.LogLavaEvent(ctx, k.Logger(ctx), types.UnstakeFromUnbond, details, "unstaking provider due to unbond that lowered its stake below min self delegation")
+			unstakeHoldBlocks := k.epochstorageKeeper.GetUnstakeHoldBlocks(ctx, stakeEntry.Chain)
+			return k.epochstorageKeeper.AppendUnstakeEntry(ctx, stakeEntry, unstakeHoldBlocks)
+		} else if stakeEntry.EffectiveStake().LT(k.specKeeper.GetMinStake(ctx, chainID).Amount) {
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/dualstaking/keeper/delegate.go:258]: done

File: x/dualstaking/keeper/hooks_test.go [hunk 6]
```diff
@@ -532,6 +532,60 @@ func TestNotRoundedShares(t *testing.T) {
...
+	for i := 0; i < 5; i++ {
+		provider, _ := ts.GetAccount(common.PROVIDER, i)
+		err := ts.StakeProvider(provider.Addr.String(), ts.spec, sdk.NewIntFromUint64(9999).Int64())
+		require.NoError(t, err)
+	}
+
+	providerAcct, provider := ts.GetAccount(common.PROVIDER, 0)
+
+	// provider completely unbond from validator, delegation is removed, but stakeentry still exists
+	_, err = ts.TxUnbondValidator(providerAcct, validator, sdk.NewInt(9999))
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/dualstaking/keeper/hooks_test.go:568]: fix

File: x/pairing/keeper/staking.go (hunk 4 above)
Comment (Yaroms) [reply x/pairing/keeper/staking.go:184]: done

File: x/dualstaking/keeper/delegate.go [hunk 7]
```diff
@@ -165,106 +165,84 @@ func (k Keeper) decreaseDelegation(ctx sdk.Context, delegator, provider, chainID
...
+				utils.Attribute{Key: "spec", Value: chainID},
+			)
 		}
+		details["min_self_delegation"] = k.GetParams(ctx).MinSelfDelegation.String()
+		utils.LogLavaEvent(ctx, k.Logger(ctx), types.UnstakeFromUnbond, details, "unstaking provider due to unbond that lowered its stake below min self delegation")
+		unstakeHoldBlocks := k.epochstorageKeeper.GetUnstakeHoldBlocks(ctx, stakeEntry.Chain)
+		return k.epochstorageKeeper.AppendUnstakeEntry(ctx, stakeEntry, unstakeHoldBlocks)
+	} else if stakeEntry.EffectiveStake().LT(k.specKeeper.GetMinStake(ctx, chainID).Amount) {
+		details["min_spec_stake"] = k.specKeeper.GetMinStake(ctx, chainID).String()
+		utils.LogLavaEvent(ctx, k.Logger(ctx), types.FreezeFromUnbond, details, "freezing provider due to unbond that lowered its stake below min spec stake")
```
Comment (omerlavanet) [inline x/dualstaking/keeper/delegate.go:242]: not necessarily unbond, can be initial stake is too low

File: x/pairing/keeper/staking.go [hunk 8]
```diff
@@ -182,13 +182,6 @@ func (k Keeper) StakeNewEntry(ctx sdk.Context, validator, creator, chainID strin
...
 
-	if stakeEntry.EffectiveStake().LT(spec.MinStakeProvider.Amount) {
```
Comment (omerlavanet) [inline x/pairing/keeper/staking.go]: can you add checks to frozen state in stake entry unit tests?

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: x/dualstaking/keeper/delegate.go (hunk 7 above)
Comment (Yaroms) [reply x/dualstaking/keeper/delegate.go]: done

File: x/pairing/keeper/staking.go (hunk 8 above)
Comment (Yaroms) [reply x/pairing/keeper/staking.go]: done

File: x/pairing/keeper/msg_server_stake_provider_test.go [hunk 9]
```diff
@@ -722,3 +723,34 @@ func TestStakeEndpoints(t *testing.T) {
 		})
 	}
 }
+
```
Comment (omerlavanet) [inline x/pairing/keeper/msg_server_stake_provider_test.go:726]: let's test that in a regular (correct) stake he is not frozen too,
let's add a test that he is frozen when staking a little when delegators add still frozen, but he can unfreeze and then unfrozen

**Review by omerlavanet — CHANGES_REQUESTED** [review]

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**